cmd/kured/main.go

```diff
@@ -725,16 +739,22 @@ func rebootAsRequired(nodeID string, booter reboot.Reboot, sentinelCommand []str
 			continue
 		}
 
+		node, err := client.CoreV1().Nodes().Get(context.TODO(), nodeID, metav1.GetOptions{})
+		if err != nil {
+			log.Fatalf("Error retrieving node object via k8s API: %v", err)
+		}
+
+		if lastSuccessfulRebootWithinMinRebootPeriod(node) {
+			log.Infof("Last successful reboot within minimal reboot period")
```

Comment:
I think it would be preferable to log the next time that a reboot will be allowed. So maybe we convert lastSuccessfulRebootWithinMinRebootPeriod() into a "getNextAllowedRebootTime()" func, and then do something like 

```golang
nextRebootTime := getNextAllowedRebootTime(node)
if time.Now().Before(nextRebootTime.Add(minRebootPeriod)) {
    log.Infof("Not allowed to reboot until %s", nextRebootTime)
}
```

Reply:
example confuses me. Why add `minRebootPeriod` to the value returned by `getNextAllowedRebootTime`?

Either way returning the next allowed reboot time (without adding the minRebootPeriod) or returning the last successful reboot time (and adding minRebootPeriod) sound good to me.

But what should we do if we don't know the last reboot time? Right now we just return "you should reboot" when there is no last successful reboot time. Would the getNextAllowedRebootTime just return `&time.Time{}` if we don't know about the last successful reboot?

Reply:
Sorry, copy/paste error in example (updated!).

In the instance where we aren't able to determine the last reboot time we can probably return time.Now(), and then update the statement to

```golang
if time.Now().Compare(nextRebootTime) < 0 {
    log.Infof("Not allowed to reboot until %s", nextRebootTime)
}
```

Reply:
are you ok with returning and error explicitly? Like 
```go 
func nextRebootTime(node) (*time.Time, error)
```
So we don't have to return time.Now() when we don't know.

Reply:
for example we could do 
```go
if minRebootPeriod > 0 {
	if t, err := nextAllowedReboot(node); err != nil {
		log.Warnf("Failed to determine next allowed reboot time: %s", err.Error())
	} else if diff := t.Sub(time.Now()); diff > 0 {
		log.Infof("Reboot not allowed until %s (%s)", t.String(), diff.String())
	}
}

```

if we don't care about logging the duration (in how much time can we reboot), we should use `time.Now().Before(t)` instead, I guess. I am really bad with subtracting times in my head.

I see that it would work without returning an error. It would even take less code, but the users would never see the warning about a missing annotation or anything.

Reply:
I'm happy with an error, but in the case of the annotation not being found, that's not actually an error, as this is a new feature and plenty of users won't configure kured to use it. So silently ignoring that and returning time.Now() seems reasonable.

Reply:
We would only check the annotation if `minRebootPeriod` is configured to be larger than 0.

Right, if you suddenly turn on the feature you would get one "wrong" error log entry before the first reboot. After the first reboot with the `--minRebootPeriod` flag, a missing annotation would actually be a real error.

If we don't care so much about logging that error, I will totally agree to returning `time.Now()` if something unexpected happens.

Reply:
> Right, if you suddenly turn on the feature you would get one "wrong" error log entry before the first reboot.

Could you clarify why this is so?

Reply:
You run cured without the `--min-reboot-period` flag. Kured will not annotate the node with the `weave.works/kured-last-successful-reboot` flag. Now you want to use the "new" feature, edit your manifest and roll out the new kured daemon set. Normally none of the pods will hold the lock from https://github.com/kubereboot/kured/blob/d0bdc115a7513721ffa310a031548bc3aea18848/cmd/kured/main.go#L666. That means they won't add the `weave.works/kured-last-successful-reboot` annotation to its nodes. We wouldn't want that anyways because then restarting/rollilng out kured daemon set would falsify the last successful reboot time. Since the feature is enabled now, kured would check for the `weave.works/kured-last-successful-reboot` annotation, but fail since it is not there yet. So if we log an error when the annotation is missing, we would see one error log for each tick until the first reboot. That the annotation is missing is expected before the first reboot. However, after the first reboot the missing annotation would be an unexpected error.

Only after the first reboot with the `--min-reboot-period` flag, this feature would actually work.

Reply:
I am not sure what is right here. Should we never log an error on missing annotations to avoid confusion before the first reboot to the cost of silencing real errors after the first reboot? Or the other way around? (I added a commit that would log the errors, happy to change it though)